apps/system/js/audio_channel_manager.js

```diff
@@ -0,0 +1,213 @@
+     * and update `_activeAudioChannels` and `_interruptedAudioChannels`.
+     *
+     * @param {AudioChannelController} audioChannel The audio channel.
+     */
+    _handleAudioChannel: function(audioChannel) {
+      audioChannel.proceedPolicy((policy) => {
+        if (policy.isAllowedToPlay) {
+          this._activeAudioChannels.set(
+            audioChannel.instanceID,
+            audioChannel
+          );
+          this.debug('Playing ' + audioChannel.instanceID);
+        } else {
+          this._activeAudioChannels.delete(audioChannel.instanceID);
+          if (policy.isNeededToResumeWhenOtherEnds) {
+            this._interruptedAudioChannels.push(audioChannel);
+            this.debug('Interrupted ' + audioChannel.instanceID);
+          }
+        }
+      });
+    },
+
+    /**
+     * Fade in all faded out audio channels.
+     */
+    _fadeInFadedOutAudioChannels: function() {
+      this._activeAudioChannels.forEach((audioChannel) => {
+        audioChannel.isFadingOut() && audioChannel
+          .setPolicy({ isNeededToFadeOut: false })
+          .proceedPolicy();
+      });
+    },
+
+    /**
+     * Resume interrupted audio channels.
+     *
+     * @param {AppWindow} [app] The app window in foreground.
+     */
+    _resumeAudioChannels: function(app) {
+      var audioChannel;
+      // Resume the app's audio channels.
+      app && app.audioChannels.forEach((audioChannel) => {
+        audioChannel.isActive() && this._manageAudioChannels(audioChannel);
+        if (audioChannel.isPlaying()) {
+          this._deleteAudioChannelFromInterruptedAudioChannels(audioChannel);
+        }
+      });
+      // Resume the latest interrupted audio channel.
+      var length = this._interruptedAudioChannels.length;
+      if (this._activeAudioChannels.size === 0 && length) {
+        audioChannel = this._interruptedAudioChannels[length - 1];
+        audioChannel.setPolicy({ isAllowedToPlay: true });
+        this._handleAudioChannel(audioChannel);
+        audioChannel.isPlaying() && this._interruptedAudioChannels.pop();
+      }
+    },
+
+    /**
+     * Delete the audio channel from `_interruptedAudioChannels` array.
+     *
+     * @param {AudioChannelController} audioChannel
+     * The audio channel want to delete.
+     */
+    _deleteAudioChannelFromInterruptedAudioChannels: function(audioChannel) {
+      var index = this._interruptedAudioChannels
+        .findIndex(function(interruptedAudioChannel) {
+          return interruptedAudioChannel.instanceID ===
+            audioChannel.instanceID;
+      });
+      index !== -1 && this._interruptedAudioChannels.splice(index, 1);
+    },
+
+    /**
+     * Check the audio channel is in background or not.
+     *
+     * @param {AudioChannelController} audioChannel The audio channel.
+     * @retrun {Boolean}
+     */
+    _isAudioChannelInBackground: function(audioChannel) {
+      return this._topMostWindow.instanceID !== audioChannel.app.instanceID;
```

Comment:
I do think you need some error handling here..what if top most window is not changed? where is your this._topMostWindow being set?